core/src/main/scala/org/apache/spark/MapOutputTracker.scala

```diff
@@ -162,7 +162,7 @@ private class ShuffleStatus(numPartitions: Int) extends Logging {
    */
   def removeOutputsOnHost(host: String): Unit = withWriteLock {
     logDebug(s"Removing outputs for host ${host}")
-    removeOutputsByFilter(x => x.asInstanceOf[BlockManagerId].host == host)
+    removeOutputsByFilter(x => x.asInstanceOf[HostLocation].host == host)
```

Comment:
Could we do `x.isInstanceOf[HostLocation] && x.asInstanceOf[HostLocation].host == host`, since people could have `Location` implementation which does not implement `HostLocation`?

Reply:
The function is used for host-based locations intentionally. So, ideally, the caller should be responsible for it.

I can add an assertion like `assert(x.isInstanceOf[HostLocation], "Location is not HostLocation")` for the safety,

Reply:
+1 for adding assertion.

Reply:
`removeOutputsOnHost`/`removeOutputsOnExecutor` may be called by other Spark internal code and the caller may not know or cannot make sure the underlying `location` is `HostLocation`/`ExecutorLocation`. 

For example in Remote/Disaggregated Shuffle Service, the location could be a custom class (not `HostLocation`/`ExecutorLocation`) since the shuffle data is on another location (shuffle server or remote storage). When there is shuffle `FetchFailed` failure, such failure will trigger code to call `removeOutputsOnHost`/`removeOutputsOnExecutor`. Then `asInstanceOf[...]` here will throw exception.

Adding `isInstanceOf[...]` in the filter predicate will make the code more generic here to support Remote/Disaggregated Shuffle Service.